relay-server/src/processing/check_ins/mod.rs

```diff
@@ -0,0 +1,143 @@
+use std::sync::Arc;
+
+use relay_quotas::{DataCategory, RateLimits};
+
+use crate::Envelope;
+use crate::envelope::{EnvelopeHeaders, Item, ItemType, Items};
+use crate::managed::{Counted, Managed, ManagedEnvelope, OutcomeError, Quantities, Rejected};
+use crate::processing::{self, Context, CountRateLimited, Forward, Output, QuotaRateLimiter};
+use crate::services::outcome::{DiscardReason, Outcome};
+
+mod process;
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// The check-ins are rate limited.
+    #[error("rate limited")]
+    RateLimited(RateLimits),
+    /// Failed to process the check-in.
+    #[error("failed to process checkin: {0}")]
+    Processing(#[from] relay_monitors::ProcessCheckInError),
+}
+
+impl OutcomeError for Error {
+    type Error = Self;
+
+    fn consume(self) -> (Option<Outcome>, Self::Error) {
+        let outcome = match &self {
+            Self::RateLimited(limits) => {
+                let reason_code = limits.longest().and_then(|limit| limit.reason_code.clone());
+                Some(Outcome::RateLimited(reason_code))
+            }
+            Self::Processing(relay_monitors::ProcessCheckInError::Json(_)) => {
+                Some(Outcome::Invalid(DiscardReason::InvalidJson))
+            }
+            Self::Processing(_) => Some(Outcome::Invalid(DiscardReason::InvalidCheckIn)),
+        };
+        (outcome, self)
+    }
+}
+
+impl From<RateLimits> for Error {
+    fn from(value: RateLimits) -> Self {
+        Self::RateLimited(value)
+    }
+}
+
+/// A processor for Check-Ins.
+pub struct CheckInsProcessor {
+    limiter: Arc<QuotaRateLimiter>,
```

Comment:
I wonder if the `limiter` is something that we could move into `Context` in the future. Hard to imagine a data type that does not need to enforce quotas.

Reply:
I wanted to keep the context 'clean' and really only have read only data related to the item being processed.

We have an example where the span processor needs Geo-IP data, for me this falls into a same category as the rate limiter, so if we were to extend the context, we should also add the geo-ip database, and all the other things we might/will need in the future. Which could be fine, but I am a bit worried it is going to explode again.

Although the point you're raising is fair, it's not the best experience to create the processor and pass in the same thing everytime. I was thinking about some kind of light dependency injection mechanism but haven't found something that I like yet.